src/ralph/models/selector.py

```diff
@@ -0,0 +1,147 @@
+"""Model selector definition"""
+
+from collections import Counter
+from dataclasses import dataclass
+from importlib import import_module
+from itertools import chain
+from typing import Any, Type, Union
+
+from pydantic import BaseModel
+
+from ralph.defaults import MODEL_PATH_SEPARATOR
+from ralph.exceptions import ModelRulesException
+from ralph.utils import get_dict_value_from_path
+
+
+@dataclass(frozen=True)
+class LazyModelField:
+    """Represents a model field."""
+
+    path: tuple[str]
+
+    def __init__(self, path: str):
+        object.__setattr__(self, "path", tuple(path.split(MODEL_PATH_SEPARATOR)))
+
+
+@dataclass(frozen=True)
+class Rule:
+    """Represents a rule used for selection."""
+
+    field: LazyModelField
+    value: Union[LazyModelField, Any]
+
+    def check(self, event):
+        """Checks if event matches the rule.
+
+        Args:
+            event (dict): The event to check.
+        """
+
+        event_value = get_dict_value_from_path(event, self.field.path)
+        expected_value = self.value
+        if isinstance(expected_value, LazyModelField):
+            expected_value = get_dict_value_from_path(event, expected_value.path)
+        return event_value == expected_value
+
+
+def selector(**filters):
+    """Returns a list of rules that should match in order to select an event.
+
+    Args:
+        **filters: Each keyword represents the path to a field, the corresponding
+            value represents the value which the field should match.
+            Example: `event_type="server"` will generate a Rule verifying that
+            `event.event_type == "server"`
+    """
+
+    return [Rule(LazyModelField(field), value) for field, value in filters.items()]
+
+
+class ModelSelector:
+    """Selects the matching model for a given event.
+
+    Attributes:
+        model_rules (dict): Stores the list of rules for each model.
+        decision_tree (dict): Stores the rule checking order for model selection.
+    """
+
+    def __init__(self, module="ralph.models.edx"):
+        """Instantiates ModelSelector."""
+
+        self.model_rules: dict[Type[BaseModel], list[Rule]] = {}
+        for object_ in import_module(module).__dict__.values():
+            if (
+                isinstance(object_, type)
+                and issubclass(object_, BaseModel)
+                and hasattr(object_, "__selector__")
+            ):
+                self.add_to_model_rules(object_, object_.__selector__)
+        self.decision_tree = self.get_decision_tree(self.model_rules)
+
+    def add_to_model_rules(self, new_model: Type[BaseModel], new_rules: list[Rule]):
```

Comment:
Maybe we should enforce this check by replacing model_rules with a subclass of dict with a __setter__ method ? [Example](https://stackoverflow.com/a/3137768)
WDYT ?

Reply:
Good idea.

Reply:
Thanks) Updated the PR)